### PR TITLE
Psalm Integration & Laminas Coding Standard 2

### LINKED ISSUE #3: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `psalm.xml` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require --dev vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Updates to Laminas Coding Standard 2 and adds `ext-json` to require.

Fixes #3 - Adds `vimeo/psalm` to dev dependencies and sets baseline.

In `src` nearly all changes are cosmetic with the exception of the following:

- ServerUrlHelper::__invoke() signature change from `string $path = null` to `?string $path = null`
- ServerUrlHelper::generate() signature change from `string $path = null` to `?string $path = null`
- UrlHelper::__invoke() signature change. $routeName and $fragmentIdentifier both nullable
- MalformedRequestBodyException::__construct() signature changed from `($message, Exception $previous = null)` to (string $message, ?Exception $previous = null)

In `test` I've added a trait to collect removed PHPUnit assertions for `assertAttribute*()` so they can more easily be found later on.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -29,19 +29,22 @@
     },
     "require": {
         "php": "^7.3 || ~8.0.0",
+        "ext-json": "*",
```
Comment (weierophinney) [inline composer.json]: Why is this added? Is it because of the body params `JsonStrategy`? If so, that's considered opt-in; not everyone will use it. I'd say this should instead become:

- A require-dev requirement AND
- A suggest entry

File: src/Exception/MalformedRequestBodyException.php [hunk 2]
```diff
@@ -10,11 +10,12 @@
 
 namespace Mezzio\Helper\Exception;
 
+use Exception;
 use InvalidArgumentException;
 
 class MalformedRequestBodyException extends InvalidArgumentException implements ExceptionInterface
 {
-    public function __construct($message, \Exception $previous = null)
+    public function __construct(string $message, ?Exception $previous = null)
```
Comment (weierophinney) [inline src/Exception/MalformedRequestBodyException.php]: This is a BC break (change in signature due to adding the `string` type to `$message`).

File: test/BodyParams/BodyParamsMiddlewareTest.php [hunk 3]
```diff
@@ -52,6 +51,7 @@ public function setUp(): void
         $this->body->rewind();
     }
 
+    /** @return RequestHandlerInterface|ObjectProphecy */
```
Comment (weierophinney) [inline test/BodyParams/BodyParamsMiddlewareTest.php]: In cases like this, it's good to also include:

```php
@psalm-return RequestHandlerInterface&ObjectProphecy
```

as it will reduce the number of errors Psalm reports.

(Do this throughout - I don't want to open a comment for each and every one!)

File: test/BodyParams/BodyParamsMiddlewareTest.php [hunk 4]
```diff
@@ -77,7 +78,8 @@ private function mockHandlerToNeverTrigger()
         return $handler;
     }
 
-    public function jsonProvider()
+    /** @return array<array-key, string[]> */
```
Comment (weierophinney) [inline test/BodyParams/BodyParamsMiddlewareTest.php:70]: Doesn't phpcs raise errors for this? If it does, make this a `@psalm-return` annotation instead (which is what I've done in other packages).

(Apply elsewhere as necessary; I see it in a number of locations.)

**Review by weierophinney — DISMISSED** [review]
Thank you for this!

You'll need to rebase (I merged a hotfix for 5.5.1, which adds some more tests for the `UrlHelper`, and thus causing conflicts here).

My main concerns are a couple of BC breaks, and cases where we may need to use Psalm-specific annotations.

File: src/Exception/MalformedRequestBodyException.php (hunk 2 above)
Comment (gsteel) [reply src/Exception/MalformedRequestBodyException.php]: Reverted!

File: test/BodyParams/BodyParamsMiddlewareTest.php (hunk 3 above)
Comment (gsteel) [reply test/BodyParams/BodyParamsMiddlewareTest.php]: For the return types, I dropped prophecy here in favour of PHPUnit mocks because adding the annotations caused more problems that they solved and my psalm-fu is not exactly black-belt standard

File: test/BodyParams/BodyParamsMiddlewareTest.php (hunk 4 above)
Comment (gsteel) [reply test/BodyParams/BodyParamsMiddlewareTest.php:70]: phpcs doesn't complain at all and adding a duplicate `@psalm-return` doesn't appear to affect psalm's "score"

Comment (weierophinney) [reply test/BodyParams/BodyParamsMiddlewareTest.php:70]: Good to know - phpcs used to complain, so this means current versions are recognizing this syntax.
